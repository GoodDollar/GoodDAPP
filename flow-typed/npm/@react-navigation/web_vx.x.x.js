// flow-typed signature: 6afd15d96d524b48dd70fa758373d370
// flow-typed version: <<STUB>>/@react-navigation/web_v^1.0.0-alpha.7/flow_v0.89.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@react-navigation/web'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@react-navigation/web' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@react-navigation/web/dist/createBrowserApp' {
  declare module.exports: any;
}

declare module '@react-navigation/web/dist/handleServerRequest' {
  declare module.exports: any;
}

declare module '@react-navigation/web/dist/index' {
  declare module.exports: any;
}

declare module '@react-navigation/web/dist/Link' {
  declare module.exports: any;
}

declare module '@react-navigation/web/src/__tests__/createBrowserApp-test' {
  declare module.exports: any;
}

declare module '@react-navigation/web/src/createBrowserApp' {
  declare module.exports: any;
}

declare module '@react-navigation/web/src/handleServerRequest' {
  declare module.exports: any;
}

declare module '@react-navigation/web/src/index' {
  declare module.exports: any;
}

declare module '@react-navigation/web/src/Link' {
  declare module.exports: any;
}

// Filename aliases
declare module '@react-navigation/web/dist/createBrowserApp.js' {
  declare module.exports: $Exports<'@react-navigation/web/dist/createBrowserApp'>;
}
declare module '@react-navigation/web/dist/handleServerRequest.js' {
  declare module.exports: $Exports<'@react-navigation/web/dist/handleServerRequest'>;
}
declare module '@react-navigation/web/dist/index.js' {
  declare module.exports: $Exports<'@react-navigation/web/dist/index'>;
}
declare module '@react-navigation/web/dist/Link.js' {
  declare module.exports: $Exports<'@react-navigation/web/dist/Link'>;
}
declare module '@react-navigation/web/src/__tests__/createBrowserApp-test.js' {
  declare module.exports: $Exports<'@react-navigation/web/src/__tests__/createBrowserApp-test'>;
}
declare module '@react-navigation/web/src/createBrowserApp.js' {
  declare module.exports: $Exports<'@react-navigation/web/src/createBrowserApp'>;
}
declare module '@react-navigation/web/src/handleServerRequest.js' {
  declare module.exports: $Exports<'@react-navigation/web/src/handleServerRequest'>;
}
declare module '@react-navigation/web/src/index.js' {
  declare module.exports: $Exports<'@react-navigation/web/src/index'>;
}
declare module '@react-navigation/web/src/Link.js' {
  declare module.exports: $Exports<'@react-navigation/web/src/Link'>;
}
