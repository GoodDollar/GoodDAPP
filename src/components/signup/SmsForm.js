// @flow
import React, { useEffect, useState } from 'react'
import { KeyboardAvoidingView } from 'react-native'
import { isIOS } from '../../lib/utils/platform'
import logger from '../../lib/logger/pino-logger'
import API, { getErrorMessage } from '../../lib/API/api'
import { getDesignRelativeHeight } from '../../lib/utils/sizes'
import { withStyles } from '../../lib/styles'
import SpinnerCheckMark from '../common/animations/SpinnerCheckMark'
import Section from '../common/layout/Section'
import ErrorText from '../common/form/ErrorText'
import OtpInput from '../common/form/OtpInput'
import CustomWrapper from './signUpWrapper'
import type { SignupState } from './SignupState'

const log = logger.child({ from: 'SmsForm' })

type Props = {
  phone: string,
  data: SignupState,
  doneCallback: ({ isPhoneVerified: boolean }) => null,
  screenProps: any,
}

export type SMSRecord = {
  smsValidated: boolean,
  sentSMS?: boolean,
}

type State = SMSRecord & {
  errorMessage: string,
  sendingCode: boolean,
  renderButton: boolean,
  resentCode: boolean,
  loading: boolean,
  otp: Array<string>,
}

const NumInputs: number = 6

class SmsForm extends React.Component<Props, State> {
  state = {
    smsValidated: false,
    sentSMS: false,
    errorMessage: '',
    sendingCode: false,
    showWait: true,
    resentCode: false,
    loading: false,
    otp: Array(NumInputs).fill(null),
  }

  handleChange = async (otp: array) => {
    const otpValue = otp.filter(val => val).join('')
    if (otpValue.replace(/ /g, '').length === NumInputs) {
      this.setState({
        loading: true,
        otp,
      })
      try {
        await this.verifyOTP(otpValue)
        this.handleSubmit()
      } catch (e) {
        if (e.ok === 0) {
          log.warn('Verify otp failed', e.message, e)
        } else {
          log.error('Verify otp failed', e.message, e)
        }

        this.setState({
          errorMessage: e.message || e,
          loading: false,
        })
      }
    } else {
      this.setState({
        errorMessage: '',
        otp,
      })
    }
  }

  handleSubmit = async () => {
    this.setState({ loading: false })
    await this.props.screenProps.doneCallback({ smsValidated: true })
  }

  // eslint-disable-next-line class-methods-use-this
  verifyOTP(otp: string) {
    return API.verifyMobile({ otp })
  }

  handleRetryWithCall = async () => {
    await this.handleRetry({ otpChannel: 'call' })
  }

  handleRetry = async (options: any) => {
    const { otpChannel } = options
    this.setState({ sendingCode: true, otp: Array(NumInputs).fill(null), errorMessage: '' })
    let { retryFunctionName } = this.props.screenProps

    retryFunctionName = retryFunctionName || 'sendOTP'

    try {
      await API[retryFunctionName]({ ...this.props.screenProps.data, otpChannel })
      this.setState({ sendingCode: false, resentCode: true })
    } catch (e) {
      const errorMessage = getErrorMessage(e)
      const exception = new Error(errorMessage)

      log.error('Resend sms code failed', errorMessage, exception)

      this.setState({
        errorMessage,
        sendingCode: false,
        resentCode: false,
      })
    }
  }

  render() {
    const { errorMessage, otp, renderButton, sendingCode, resentCode } = this.state
    const { styles } = this.props

    return (
      <KeyboardAvoidingView behavior={isIOS ? 'padding' : 'height'} style={styles.keyboardAvoidWrapper}>
        <CustomWrapper handleSubmit={this.handleSubmit} footerComponent={() => <React.Fragment />}>
          <Section grow justifyContent="flex-start">
            <Section.Stack justifyContent="flex-start" style={styles.container}>
              <Section.Row justifyContent="center">
                <Section.Title color="darkGray" fontSize={22} fontWeight="500" textTransform="none">
                  {'Enter the verification code\nsent to your phone'}
                </Section.Title>
              </Section.Row>
              <Section.Stack justifyContent="center" style={styles.bottomContent}>
                <OtpInput
                  shouldAutoFocus
                  numInputs={NumInputs}
                  onChange={this.handleChange}
                  hasErrored={errorMessage !== ''}
                  errorStyle={styles.errorStyle}
                  value={otp}
                  placeholder="*"
                  isInputNum={true}
                  aside={[3]}
                />
                <ErrorText error={errorMessage} />
              </Section.Stack>
            </Section.Stack>
            <Section.Row alignItems="center" justifyContent="center" style={styles.row}>
              <SMSAction
                styles={styles}
                sendingCode={sendingCode}
                resentCode={resentCode}
                renderButton={renderButton}
                handleRetry={this.handleRetry}
                handleRetryWithCall={this.handleRetryWithCall}
                onFinish={() => {
                  //reset smsaction state
                  this.setState({ resentCode: false })
                }}
              />
            </Section.Row>
          </Section>
        </CustomWrapper>
      </KeyboardAvoidingView>
    )
  }
}

const SMSAction = ({ handleRetry, handleRetryWithCall, resentCode, sendingCode, onFinish, styles }) => {
  const [showWait, setWait] = useState(true)

  useEffect(() => {
    if (showWait) {
      setTimeout(() => {
        setWait(false)
      }, 10000)
    }
  }, [showWait])

  if (showWait === false) {
    return (
      <SpinnerCheckMark
        height={100}
        loading={sendingCode}
        success={resentCode}
        onFinish={() => {
          setWait(true)
          onFinish()
        }}
      >
        <Section.Text
          textDecorationLine="underline"
          fontWeight="medium"
          fontSize={14}
          color="primary"
          onPress={handleRetry}
          style={styles.sendCodeAgainButton}
        >
          Send the code again
        </Section.Text>
        <Section.Text
          textDecorationLine="underline"
          fontWeight="medium"
          fontSize={14}
          color="primary"
          onPress={handleRetryWithCall}
        >
          Send via voice call
        </Section.Text>
      </SpinnerCheckMark>
    )
  }

  return (
    <Section.Text fontSize={14} color="gray80Percent">
      Please wait a few seconds until the SMS arrives
    </Section.Text>
  )
}

const getStylesFromProps = ({ theme }) => ({
  keyboardAvoidWrapper: {
    width: '100%',
  },
  informativeParagraph: {
    margin: '1em',
  },
  buttonWrapper: {
    alignContent: 'stretch',
    flexDirection: 'column',
    display: 'flex',
    justifyContent: 'space-between',
  },
  button: {
    justifyContent: 'center',
    width: '100%',
    height: 60,
  },
  row: {
    marginVertical: theme.sizes.defaultDouble,
  },
  errorStyle: {
    borderBottomWidth: 1,
    borderBottomColor: theme.colors.red,
    color: theme.colors.red,
  },
  container: {
    minHeight: getDesignRelativeHeight(200),
    height: getDesignRelativeHeight(200),
  },
  bottomContent: {
    marginTop: 'auto',
    marginBottom: theme.sizes.defaultDouble,
  },
  sendCodeAgainButton: {
    marginRight: 'auto',
  },
})

export default withStyles(getStylesFromProps)(SmsForm)
