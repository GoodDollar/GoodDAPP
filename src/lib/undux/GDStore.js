// @flow
import { createConnectedStore, type StoreDefinition } from 'undux'
import { compose } from 'lodash/fp'
import effects from '../../lib/undux/effects'
import type { StandardFeed } from '../gundb/StandardFeed'
import withPinoLogger from './plugins/logger'
import { createUseCurriedSettersHook } from './utils/setter'

/**
 * Account data
 * @type {{balance: ?string}, {entitlement: ?string}, {ready: false}}
 */
type Account = {
  balance: ?string,
  entitlement: ?string,
  ready: false,
}

export type VerificationState = {
  attemptsCount: number,
  attemptsHistory: string[],
  reachedMaxAttempts: boolean,
}

/**
 * Type definition for the global store
 * @type {
    {balanceUpdate: BalanceUpdate},
    {account: Account},
    {destinationPath: string},
    {feeds: StandardFeed[]},
    {feedLoading: Boolean}
  }
 */
export type State = {
  balanceUpdate: boolean,
  account: Account,
  destinationPath: string,
  feeds: StandardFeed[],
  feedLoading: boolean,
  currentFeed: any,
  verification: VerificationState,
}

export const defaultVerificationState: VerificationState = {
  attemptsCount: 0,
  attemptsHistory: [],
  reachedMaxAttempts: false,
}

/**
 * Initial store state
 * @constant
 */
const initialState: State = {
  balanceUpdate: false,
  account: {
    balance: undefined,
    entitlement: undefined,
    ready: false,
  },
  verification: {
    ...defaultVerificationState,
  },
  isLoggedInCitizen: false,
  isLoggedIn: false,
  profile: {},
  privateProfile: {},
  destinationPath: '',
  feeds: [],
  currentFeed: undefined,
  feedLoading: false,
  inviteCode: undefined,
}

/**
 * default exported instance of our global Undux Store
 * @module
 */
const GDStore: StoreDefinition<State> = createConnectedStore(
  initialState,
  compose(
    effects,
    withPinoLogger,
  ),
)

export const useCurriedSetters = createUseCurriedSettersHook(() => GDStore)

export default GDStore
