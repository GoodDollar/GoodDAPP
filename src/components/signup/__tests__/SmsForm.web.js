import { createSwitchNavigator } from '@react-navigation/core'
import { createBrowserApp } from '@react-navigation/web'
import React from 'react'
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'

import GDStore from '../../../lib/undux/GDStore'
import SmsForm from '../SmsForm.web'

const { Container } = GDStore

describe('SmsForm', () => {
  it('renders without errors', () => {
    const WebRouter = createBrowserApp(createSwitchNavigator({ SmsForm }))
    const tree = renderer.create(
      <Container>
        <WebRouter />
      </Container>
    )
    expect(tree.toJSON()).toBeTruthy()
  })

  it('matches snapshot', () => {
    const WebRouter = createBrowserApp(createSwitchNavigator({ SmsForm }))
    const component = renderer.create(
      <Container>
        <WebRouter />
      </Container>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
