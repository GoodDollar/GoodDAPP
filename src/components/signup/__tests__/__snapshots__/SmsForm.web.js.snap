// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmsForm matches snapshot 1`] = `
<div
  className="css-view-1dbjc4n"
  style={
    Object {
      "WebkitAlignItems": "stretch",
      "WebkitBoxAlign": "stretch",
      "WebkitBoxDirection": "normal",
      "WebkitBoxFlex": 1,
      "WebkitBoxOrient": "vertical",
      "WebkitBoxPack": "justify",
      "WebkitFlexBasis": "0%",
      "WebkitFlexDirection": "column",
      "WebkitFlexGrow": 1,
      "WebkitFlexShrink": 1,
      "WebkitJustifyContent": "space-between",
      "alignItems": "stretch",
      "display": "flex",
      "flexBasis": "0%",
      "flexDirection": "column",
      "flexGrow": 1,
      "flexShrink": 1,
      "justifyContent": "space-between",
      "maxWidth": "500px",
      "msFlexAlign": "stretch",
      "msFlexDirection": "column",
      "msFlexNegative": 1,
      "msFlexPack": "justify",
      "msFlexPositive": 1,
      "msFlexPreferredSize": "0%",
      "paddingBottom": "20px",
      "paddingLeft": "20px",
      "paddingRight": "20px",
      "paddingTop": "20px",
    }
  }
>
  <div
    className="css-view-1dbjc4n"
    style={
      Object {
        "WebkitBoxFlex": 1,
        "WebkitBoxPack": "space-evenly",
        "WebkitFlexBasis": "0%",
        "WebkitFlexGrow": 1,
        "WebkitFlexShrink": 1,
        "WebkitJustifyContent": "space-evenly",
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "space-evenly",
        "msFlexNegative": 1,
        "msFlexPack": "space-evenly",
        "msFlexPositive": 1,
        "msFlexPreferredSize": "0%",
        "paddingTop": "30px",
      }
    }
  >
    <div
      className="css-text-76zvg2"
      dir="auto"
      style={
        Object {
          "color": "rgba(85,85,85,1.00)",
          "direction": "ltr",
          "fontFamily": "Roboto, \\"Helvetica Neue\\", Helvetica, Arial, sans-serif",
          "fontSize": "24px",
          "marginBottom": "30px",
          "textAlign": "center",
        }
      }
    >
      Enter the verification code 
 sent to your phone
    </div>
    <div
      className={false}
      style={
        Object {
          "display": "flex",
          "justifyContent": "space-evenly",
        }
      }
    >
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
          }
        }
      >
        <input
          className=""
          disabled={false}
          maxLength="1"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          style={
            Object {
              "borderBottom": "1px solid #555",
              "borderLeft": "none",
              "borderRight": "none",
              "borderTop": "none",
              "fontSize": "1.5rem",
              "height": "3rem",
              "margin": "0 0.5rem",
              "textAlign": "center",
              "width": "100%",
            }
          }
          type="tel"
        />
      </div>
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
          }
        }
      >
        <input
          className=""
          disabled={false}
          maxLength="1"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          style={
            Object {
              "borderBottom": "1px solid #555",
              "borderLeft": "none",
              "borderRight": "none",
              "borderTop": "none",
              "fontSize": "1.5rem",
              "height": "3rem",
              "margin": "0 0.5rem",
              "textAlign": "center",
              "width": "100%",
            }
          }
          type="tel"
        />
      </div>
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
          }
        }
      >
        <input
          className=""
          disabled={false}
          maxLength="1"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          style={
            Object {
              "borderBottom": "1px solid #555",
              "borderLeft": "none",
              "borderRight": "none",
              "borderTop": "none",
              "fontSize": "1.5rem",
              "height": "3rem",
              "margin": "0 0.5rem",
              "textAlign": "center",
              "width": "100%",
            }
          }
          type="tel"
        />
      </div>
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
          }
        }
      >
        <input
          className=""
          disabled={false}
          maxLength="1"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          style={
            Object {
              "borderBottom": "1px solid #555",
              "borderLeft": "none",
              "borderRight": "none",
              "borderTop": "none",
              "fontSize": "1.5rem",
              "height": "3rem",
              "margin": "0 0.5rem",
              "textAlign": "center",
              "width": "100%",
            }
          }
          type="tel"
        />
      </div>
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
          }
        }
      >
        <input
          className=""
          disabled={false}
          maxLength="1"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          style={
            Object {
              "borderBottom": "1px solid #555",
              "borderLeft": "none",
              "borderRight": "none",
              "borderTop": "none",
              "fontSize": "1.5rem",
              "height": "3rem",
              "margin": "0 0.5rem",
              "textAlign": "center",
              "width": "100%",
            }
          }
          type="tel"
        />
      </div>
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
          }
        }
      >
        <input
          className=""
          disabled={false}
          maxLength="1"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          style={
            Object {
              "borderBottom": "1px solid #555",
              "borderLeft": "none",
              "borderRight": "none",
              "borderTop": "none",
              "fontSize": "1.5rem",
              "height": "3rem",
              "margin": "0 0.5rem",
              "textAlign": "center",
              "width": "100%",
            }
          }
          type="tel"
        />
      </div>
    </div>
    <div
      className="css-view-1dbjc4n"
    >
      <div
        className="css-text-76zvg2"
        dir="auto"
        style={
          Object {
            "color": "rgba(255,0,0,1.00)",
            "direction": "ltr",
            "fontFamily": "Roboto, \\"Helvetica Neue\\", Helvetica, Arial, sans-serif",
            "fontSize": "18px",
            "marginBottom": "30px",
            "marginTop": "30px",
            "minHeight": "24px",
            "textAlign": "center",
          }
        }
      />
    </div>
    <div
      className="css-view-1dbjc4n"
      style={
        Object {
          "WebkitAlignContent": "stretch",
          "WebkitBoxDirection": "normal",
          "WebkitBoxOrient": "vertical",
          "WebkitBoxPack": "justify",
          "WebkitFlexDirection": "column",
          "WebkitJustifyContent": "space-between",
          "alignContent": "stretch",
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "space-between",
          "msFlexDirection": "column",
          "msFlexLinePack": "stretch",
          "msFlexPack": "justify",
        }
      }
    >
      <div
        className="css-text-76zvg2"
        dir="auto"
        style={
          Object {
            "color": "rgba(85,85,85,1.00)",
            "direction": "ltr",
            "fontFamily": "Roboto, \\"Helvetica Neue\\", Helvetica, Arial, sans-serif",
            "fontSize": "18px",
            "marginTop": "30px",
            "textAlign": "center",
          }
        }
      >
        Please wait a few seconds until the SMS arrives
      </div>
    </div>
  </div>
  <div
    className="css-view-1dbjc4n"
    style={
      Object {
        "WebkitBoxFlex": 1,
        "WebkitBoxPack": "end",
        "WebkitFlexBasis": "0%",
        "WebkitFlexGrow": 1,
        "WebkitFlexShrink": 1,
        "WebkitJustifyContent": "flex-end",
        "display": "flex",
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "flex-end",
        "msFlexNegative": 1,
        "msFlexPack": "end",
        "msFlexPositive": 1,
        "msFlexPreferredSize": "0%",
        "paddingTop": "20px",
      }
    }
  />
</div>
`;
